### PR TITLE
Fix panic caused by rand_bytes()

### PR DESCRIPTION
**Pull request by vibharya**
Signed-off-by: Vibha Acharya <vibharya@amazon.co.uk>

### Summary of the PR

rand_bytes() causes a panic when number of bytes requested are not multiples of 4.
`'main' panicked at 'attempt to subtract with overflow'`

Relates to : #178

The unit test was testing for 8 which is why it passed.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: src/rand.rs [hunk 1]
```diff
@@ -156,7 +156,7 @@ mod tests {
 
     #[test]
     fn test_rand_bytes() {
-        let s = rand_bytes(8);
-        assert_eq!(8, s.len());
+        let s = rand_bytes(7);
```
Comment (andreeaflorescu) [inline src/rand.rs:159]: Let's add multiple test scenarios here. Can you instead do a for from 0 to 8 and check that rand_bytes(i).len() is i?

**Review by andreeaflorescu — CHANGES_REQUESTED** [review]
Please update the changelog to specify the bug you fixed. This is now part of the requirements for a PR.

**Comment by andreeaflorescu** [comment]
Can you also add the following to your commit message:
```Fixes: https://github.com/rust-vmm/vmm-sys-util/issues/178```

File: src/rand.rs (hunk 1 above)
Comment (vibharya) [reply src/rand.rs:159]: Absolutely!

**Comment by andreeaflorescu** [comment]
> Please update the changelog to specify the bug you fixed. This is now part of the requirements for a PR.

There is still the changelog update needed.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,11 @@
 # Changelog
+## Unreleased
+
+### Added
+
+### Changed
+- [[#178](https://github.com/rust-vmm/vmm-sys-util/issues/178)]: Fixed the
```
Comment (andreeaflorescu) [inline CHANGELOG.md:7]: Let's be more explicit about what was happening: fixed a bug in `rand_bytes` that was triggering a panic when the number of bytes was not a multiple of 4.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,4 +1,11 @@
 # Changelog
+## Unreleased
+
+### Added
```
Comment (andreeaflorescu) [inline CHANGELOG.md:4]: No need to add this as an empty section. When we're doing the release ideally the only change needed should be to replace Unreleased -> v0.x.y

**Review by andreeaflorescu — APPROVED** [review]

**Review by russell-islam — APPROVED** [review]

File: src/rand.rs [hunk 4]
```diff
@@ -81,10 +81,10 @@ fn rand_alphanumerics_impl(rand_fn: &dyn Fn() -> u32, len: usize) -> OsString {
 
 fn rand_bytes_impl(rand_fn: &dyn Fn() -> u32, len: usize) -> Vec<u8> {
     let mut buf: Vec<Vec<u8>> = Vec::new();
-    let mut num = len;
+    let mut num = if len % 4 == 0 { len / 4 } else { len / 4 + 1 };
```
Comment (jiangliu) [inline src/rand.rs:84]: How about 
```
let mut num. = (len + 3) >> 2;
```

Comment (andreeaflorescu) [reply src/rand.rs:84]: I find the if to be a bit easier to read 😬

**Review by aghecenco — APPROVED** [review]

File: src/rand.rs (hunk 4 above)
Comment (arunpkm) [reply src/rand.rs:84]: Performance is more important than readability. So, `let mut num. = (len + 3) >> 2;` should have been opted. 

Refer Perplexity AI thread: https://www.perplexity.ai/search/what-is-the-compute-time-diffe-VJ_hoMTVT92_RlvS5Jd9sg 

```
Key Observations
Option 2 is consistently faster: For all tested inputs, the bit-shifting approach (len + 3) >> 2 outperforms the conditional approach if len % 4 == 0 { len / 4 } else { len / 4 + 1 }.

Performance improvement: Option 2 is approximately 23-35% faster than Option 1, with an average improvement of about 27%.

Consistency across inputs: The performance advantage holds regardless of whether the input is divisible by 4 or not.
```

**Pull request merged**